### PR TITLE
Material 3 common buttons should use sparkle splash effect on Android.

### PR DESCRIPTION
**Pull request by darrenaustin**
Part of: #91605

Part of the Material 3 migration for the common buttons is to use the sparkle splash effect if running on Android. This will be handled by the full button migration in #100794, but that is delayed due to an API issue we need to resolve. In the mean time this PR just turns on the sparkle effect (#99731) for buttons when `ThemeData.useMaterial3` is turned on and the platform is non-web Android.

Added tests to check that the right splash factory is used for each button.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/test/material/elevated_button_test.dart [hunk 1]
```diff
@@ -1272,6 +1273,55 @@ void main() {
...
+    ));
+
+    if (debugDefaultTargetPlatformOverride! == TargetPlatform.android && !kIsWeb) {
+      expect(buttonInkWell.splashFactory, equals(InkSparkle.splashFactory));
+    } else {
+      expect(buttonInkWell.splashFactory, equals(InkSplash.splashFactory));
+    }
+  }, variant: TargetPlatformVariant.all());
+
+  testWidgets('ElevatedButton uses InkSparkle only for Android non-web when useMaterial3 is false', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/material/elevated_button_test.dart:1303]: I think this text is wrong, nothing should use InkSparkle when useMaterial3 is false, right?

File: packages/flutter/test/material/outlined_button_test.dart [hunk 2]
```diff
@@ -1445,6 +1446,55 @@ void main() {
...
+    ));
+
+    if (debugDefaultTargetPlatformOverride! == TargetPlatform.android && !kIsWeb) {
+      expect(buttonInkWell.splashFactory, equals(InkSparkle.splashFactory));
+    } else {
+      expect(buttonInkWell.splashFactory, equals(InkSplash.splashFactory));
+    }
+  }, variant: TargetPlatformVariant.all());
+
+  testWidgets('OutlinedButton uses InkSparkle only for Android non-web when useMaterial3 is false', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/material/outlined_button_test.dart:1476]: Same for this text.

File: packages/flutter/test/material/text_button_test.dart [hunk 3]
```diff
@@ -1238,6 +1239,55 @@ void main() {
...
+    ));
+
+    if (debugDefaultTargetPlatformOverride! == TargetPlatform.android && !kIsWeb) {
+      expect(buttonInkWell.splashFactory, equals(InkSparkle.splashFactory));
+    } else {
+      expect(buttonInkWell.splashFactory, equals(InkSplash.splashFactory));
+    }
+  }, variant: TargetPlatformVariant.all());
+
+  testWidgets('TextButton uses InkSparkle only for Android non-web when useMaterial3 is false', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/material/text_button_test.dart:1269]: Same for this text.

**Review by justinmc — APPROVED** [review]
✨  LGTM ✨ 

I need to try this out sometime, I still haven't seen Flutter actually do the sparkle live.  Exciting.

Just one potential problem with the test text, and one question:

So if I've got an app that builds to both iOS and Android that uses Material buttons on both, and I set useMaterial3 to true, it will not sparkle on iOS?  Just out of curiosity, is it not possible to get iOS to sparkle even if I want to?

File: packages/flutter/test/material/elevated_button_test.dart (hunk 1 above)
Comment (darrenaustin) [reply packages/flutter/test/material/elevated_button_test.dart:1303]: Le d'oh! Thx for catching that.

**Comment by darrenaustin** [comment]
> So if I've got an app that builds to both iOS and Android that uses Material buttons on both, and I set useMaterial3 to true, it will not sparkle on iOS? Just out of curiosity, is it not possible to get iOS to sparkle even if I want to?

It is intended only for Android devices so that is why we have the defaults set up that way. I don't know enough about the implementation whether it would work on iOS. @clocksmith would know more about that.

**Comment by clocksmith** [comment]
> > So if I've got an app that builds to both iOS and Android that uses Material buttons on both, and I set useMaterial3 to true, it will not sparkle on iOS? Just out of curiosity, is it not possible to get iOS to sparkle even if I want to?
> 
> It is intended only for Android devices so that is why we have the defaults set up that way. I don't know enough about the implementation whether it would work on iOS. @clocksmith would know more about that.

It will currently work on any platform that uses the non-web engine (try manually setting splashFactory for theme to InkSparkle), and eventually it will also work on the CanvasKit engine!

**Comment by clocksmith** [comment]
But yes, it is correct that the Material guidance is to only use on Android. I will be following up with a post on making custom ripples with the same tech

**Pull request merged**